lang/cs/Org.Apache.REEF.Network/Elastic/Failures/Default/DefaultFailureStateMachine.cs

```diff
@@ -0,0 +1,320 @@
+        /// a target failure state. The assumption is that higher failure states
+        /// have higher thresholds.
+        /// </summary>
+        /// <param name="level">The failure state we want to change</param>
+        /// <param name="threshold">A [0, 1] value specifying when the failure level is reached</param>
+        public void SetThreshold(IFailureState level, float threshold)
+        {
+            if (!(level is DefaultFailureState))
+            {
+                throw new ArgumentException(level.GetType() + " is not DefaultFailureStateMachine");
+            }
+
+            if (level.FailureState == (int)DefaultFailureStates.Continue)
+            {
+                throw new ArgumentException("Cannot change the threshold for Continue state");
+            }
+
+            lock (_statusLock)
+            {
+                transitionWeights[(DefaultFailureStates)level.FailureState] = threshold;
+
+                CheckConsistency();
+            }
+        }
+
+        /// <summary>
+        /// A utility method for setting multiple threshold at once.
+        /// </summary>
+        /// <param name="weights">Pairs of failure states with related new thresholds</param>
+        public void SetThresholds(params Tuple<IFailureState, float>[] weights)
+        {
+            if (!weights.All(weight => weight.Item1 is DefaultFailureState))
+            {
+                throw new ArgumentException("Input is not of type DefaultFailureStateMachine,");
+            }
+
+            if (weights.Any(weight => weight.Item1.FailureState == (int)DefaultFailureStates.Continue))
+            {
+                throw new ArgumentException("Cannot change the threshold for Continue state.");
+            }
+
+            lock (_statusLock)
+            {
+                foreach (Tuple<IFailureState, float> weight in weights)
+                {
+                    transitionWeights[(DefaultFailureStates)weight.Item1.FailureState] = weight.Item2;
+                }
+
+                CheckConsistency();
+            }
+        }
+
+        /// <summary>
+        /// Utility method used to clone the target failure machine.
+        /// Only the thresholds are cloned, while the machine state is not.
+        /// </summary>
+        /// <param name="initalPoints">How many data points are avaialble in the new state machine</param>
+        /// <param name="initalState">The state from which the new machine should start</param>
+        /// <returns>A new failure machine with the same settings</returns>
+        public IFailureStateMachine Clone(
+            int initalPoints = 0, 
+            int initalState = (int)DefaultFailureStates.Continue)
+        {
+            var newMachine = new DefaultFailureStateMachine(initalPoints, (DefaultFailureStates)initalState);
+
+            foreach (DefaultFailureStates state in transitionWeights.Keys.OrderByDescending(x => x))
```

Comment:
maybe make `transitionWeights` a `SortedDictionary`? Then we won't need that `.OrderByDescending()` call at all:
```C#
foreach (var state in transitionWeights.Keys)
```
etc.

Reply:
good point. Fixed.